### PR TITLE
Feature/binary download

### PR DESCRIPTION
**Pull request by glabelderp**
If the environment variable FIRMWARECI_BINURL is set
contest-client will download the file with the url from
the environment variable.

### CODE REVIEW

**Review by llogen — DISMISSED** [review]
LGTM

File: cli/verbs.go [hunk 1]
```diff
@@ -35,6 +36,32 @@ func run(requestor string, transport transport.Transport, stdout io.Writer) erro
...
+		//trusted domain
+		if binaryURL != "" {
+			//the file name is hardcoded atm
+			file, err := os.Create("/root/assets/TEST_BIN")
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+
+			resp, err := http.Get(binaryURL)
```
Comment (walterchris) [inline cli/verbs.go:54]: This assumes it's always http - correct? IIRC we stated multiple download methods?

File: cli/verbs.go [hunk 2]
```diff
@@ -35,6 +37,46 @@ func run(requestor string, transport transport.Transport, stdout io.Writer) erro
...
+			file, err := os.Create("/root/assets/TEST_BIN")
+			if err != nil {
+				return err
+			}
+
+			//the file name is hardcoded atm
+			defer file.Close()
+
+			switch URL.Scheme {
+			case "http":
```
Comment (walterchris) [inline cli/verbs.go:63]: The scheme can not be https? is https supported? (it should be :) )

File: cli/verbs.go [hunk 3]
```diff
@@ -35,6 +37,46 @@ func run(requestor string, transport transport.Transport, stdout io.Writer) erro
...
+		//the url is presumed to be from a
+		//trusted domain
+		if unparsedURL != "" {
+
+			URL, err := url.Parse(unparsedURL)
+			if err != nil {
+				return err
+			}
+
+			file, err := os.Create("/root/assets/TEST_BIN")
```
Comment (walterchris) [inline cli/verbs.go:54]: We do not get the filename from the URL? Or at least from the response?

File: cli/verbs.go (hunk 2 above)
Comment (glabelderp) [reply cli/verbs.go:63]: It can, will fix that

File: cli/verbs.go (hunk 3 above)
Comment (glabelderp) [reply cli/verbs.go:54]: ok, yeah we  just need to negotiate that with the test e.g. by using templating

**Comment by walterchris** [comment]
This belongs to #4

File: pkg/download/download.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+		User:            URL.User.Username(),
+		Auth:            AuthMethods,
+		HostKeyCallback: ssh.FixedHostKey(hostPublicKey),
+	})
+	if err != nil {
+		return err
+	}
+	defer sshClient.Close()
+
+	{
```
Comment (llogen) [inline pkg/download/download.go:124]: Are the curly brackets needed here?

**Review by llogen — APPROVED** [review]
Besides that one comment. LGTM

File: pkg/download/download.go (hunk 4 above)
Comment (glabelderp) [reply pkg/download/download.go:124]: I just tested this and the answer is very likely no :D

File: pkg/env/env.go [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+const (
+	DefaultBinaryDir          = "/root/assets/testbin/"
+	ENV_BIN_URL               = "FIRMWARECI_BINURL"
+	ENV_BIN_DIR               = "FIRMWARECI_BINDIR"
+	ENV_SSH_HOST_PUBLIC       = "FIRMWARECI_HOSTPUB"
+	ENV_SSH_PRIVATE_PW        = "FIRMWARECI_SSHPRIVATEPW"
+	ENV_S3_ACCESS             = "FIRMWARECI_S3_ACCESS"
+	ENV_S3_SECRET             = "FIRMWARECI_S3_SECRET"
+	ENV_S3_REGION             = "FIRMWARECI_S3_REGION"
+	FIRMWARECI_SSH_PUBLIC_KEY = "/root/.shh/ci.pub"
```
Comment (walterchris) [inline pkg/env/env.go:12]: Typo

File: pkg/download/download.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+	file, err := os.Create(binaryPath)
+	if err != nil {
+		return "", err
+	}
+
+	defer file.Close()
+
+	switch URL.Scheme {
+	case "http", "https", "":
+		{
```
Comment (walterchris) [inline pkg/download/download.go:52]: These are actually not needed.

File: pkg/download/download.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	connection, err := ftp.DialTimeout(URL.Hostname(), 60*time.Second)
+	if err != nil {
+		return err
+	}
+
+	defer connection.Quit()
+
+	pw, set := URL.User.Password()
```
Comment (walterchris) [inline pkg/download/download.go:109]: It's just a nit: I would check User and Password first, and then connect. But that's personal preference here.

File: pkg/env/env.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+const (
+	DefaultBinaryDir       = "/root/assets/testbin/"
+	EnvBinUrl              = "FIRMWARECI_BINURL"
+	EnvBinDir              = "FIRMWARECI_BINDIR"
+	EnvSSHHostPublic       = "FIRMWARECI_HOSTPUB"
+	EnvSSHPrivatePW        = "FIRMWARECI_SSHPRIVATEPW"
+	EnvS3Access            = "FIRMWARECI_S3_ACCESS"
+	EnvS3Secret            = "FIRMWARECI_S3_SECRET"
+	EnvS3Region            = "FIRMWARECI_S3_REGION"
+	FirmwareciSSHPublicKey = "/root/.shh/firmwareci.pub"
```
Comment (walterchris) [inline pkg/env/env.go:12]: still a type here. It's ssh and not shh

Comment (glabelderp) [reply pkg/env/env.go:12]: Rip, I thought I had fixed that . Sorry about that...

File: pkg/download/download.go (hunk 6 above)
Comment (glabelderp) [reply pkg/download/download.go:52]: I think they make this whole section a bit more readable and structured. Do you want me to delete them?

File: pkg/download/download.go (hunk 7 above)
Comment (glabelderp) [reply pkg/download/download.go]: Sure

**Review by llogen — COMMENTED** [review]
LGTM

File: pkg/download/download.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+	defer sshClient.Close()
+
+	sftpClient, err := sftp.NewClient(sshClient)
+	if err != nil {
+		return err
+	}
+
+	defer sftpClient.Close()
+
+	binFile, err := sftpClient.Open(URL.Path)
```
Comment (walterchris) [inline pkg/download/download.go:181]: we do not check this error.

File: pkg/download/download.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+	out, err := S3client.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: Bucket,
+		Key:    Key,
+	})
+	if err != nil {
+		return err
+	}
+
+	io.Copy(file, out.Body)
```
Comment (walterchris) [inline pkg/download/download.go:230]: io.Copy returns (int64, err) - we should check this, at least the err.

File: pkg/download/download.go [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+	sftpClient, err := sftp.NewClient(sshClient)
+	if err != nil {
+		return err
+	}
+
+	defer sftpClient.Close()
+
+	binFile, err := sftpClient.Open(URL.Path)
+
+	io.Copy(file, binFile)
```
Comment (walterchris) [inline pkg/download/download.go:183]: Look below.

File: pkg/download/download.go [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+
+	binFile.Close()
```
Comment (walterchris) [inline pkg/download/download.go:185]: defer?

File: pkg/download/download.go [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+		Bucket: Bucket,
+		Key:    Key,
+	})
+	if err != nil {
+		return err
+	}
+
+	io.Copy(file, out.Body)
+
+	out.Body.Close()
```
Comment (walterchris) [inline pkg/download/download.go:232]: defer

File: pkg/download/download.go [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+
+	resp, err := connection.Retr(URL.Path)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(file, resp); err != nil {
+		return err
+	}
+	resp.Close()
```
Comment (walterchris) [inline pkg/download/download.go:119]: defer?

**Review by walterchris — CHANGES_REQUESTED** [review]

**Review by walterchris — CHANGES_REQUESTED** [review]

File: pkg/download/download.go [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+	binFile, err := sftpClient.Open(URL.Path)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(file, binFile); err != nil {
+		return err
+	}
+
+	defer binFile.Close()
```
[the remark this answers is not in the document]
Comment (glabelderp) [reply pkg/download/download.go:190]: Oh yeah, absolutely

**Review by walterchris — APPROVED** [review]

**Pull request merged**